### PR TITLE
PHP runthrough

### PR DESCRIPTION
**Pull request by moorscode**
Simplified the class structure.
Split up the dependencies in separate classes.
Added translation file (.pot)
Renamed `frontend` to `assets`.
Introduced Facade to avoid having to use a singleton.
Convert the String Store to class and use it instead of duplicate functional implementations.
Added dependency tests.

### CODE REVIEW

File: inc/configuration/class-yoast-acf-analysis-configuration.php [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+		return get_option( 'acf_version' );
+	}
+
+	/**
+	 * @return Yoast_ACF_Analysis_String_Store
+	 */
+	public function blacklist() {
+
+		$blacklist = apply_filters(
+			Yoast_ACF_Analysis_Facade::get_filter_name( 'blacklist' ),
```
Comment (kraftner) [inline inc/configuration/class-yoast-acf-analysis-configuration.php]: Just out of curiosity interested in why you've chosen to use strings here instead of `__FUNCTION__` like I've been using in the earlier implementation.

File: inc/class-yoast-acf-analysis-registry.php [hunk 2]
```diff
@@ -13,11 +13,11 @@ class Yoast_ACF_Analysis_Registry {
...
 
 	/**
-	 * @param string|int                       $id    Registry index.
-	 * @param Yoast_ACF_Analysis_Configuration $class Registry value.
+	 * @param string|int $id   Registry index.
+	 * @param mixed      $item Item to store in the registry.
 	 */
-	public function add( $id, Yoast_ACF_Analysis_Configuration $class ) {
-		$this->storage[ $id ] = $class;
+	public function add( $id, $item ) {
```
Comment (kraftner) [inline inc/class-yoast-acf-analysis-registry.php]: While this makes sense just looking at this class you can now put *anything* in the registry because [`boot()`](https://github.com/Yoast/yoast-acf-analysis/blob/7e1de670bcd76410d48fae3ef081b6a325687fed/inc/class-yoast-acf-analysis.php#L46-L58) "trusts" the registry.
Admittedly this should probably stay like this but needs to be fixed in `boot()`with and additional check like

`$registry->get( 'config' ) instanceof Yoast_ACF_Analysis_Configuration`

File: yoast-acf-analysis.php [hunk 3]
```diff
@@ -20,8 +20,7 @@
 	define( 'YOAST_ACF_ANALYSIS_FILE', __FILE__ );
 }
 
-if ( file_exists( dirname( YOAST_ACF_ANALYSIS_FILE ) . '/vendor/autoload_52.php' ) ) {
-
+if ( is_file( dirname( YOAST_ACF_ANALYSIS_FILE ) . '/vendor/autoload_52.php' ) ) {
```
Comment (moorscode) [inline yoast-acf-analysis.php]: It's one of my pet-peeves I guess, had some issues with it in the past :)

File: inc/class-yoast-acf-analysis-requirements.php [hunk 4]
```diff
@@ -6,69 +6,40 @@
...
-					if ( version_compare( substr( WPSEO_VERSION, 0, 3 ), self::MIN_WPSEO_VERSION, '<' ) ) {
-						add_action( 'admin_notices', array( $this, 'yoast_seo_requirements_not_met' ) );
-						$deactivate = true;
-					}
-				}
-			}
-
-			// Deactivate when we cannot do the job we are hired to do.
-			if ( $deactivate ) {
-				deactivate_plugins( plugin_basename( YOAST_ACF_ANALYSIS_FILE ) );
```
Comment (moorscode) [inline inc/class-yoast-acf-analysis-requirements.php]: When I was testing the functionality, I noticed that the message would only be there for one page-load. I understand the concept of deactivating the plugin, but a user will generally assume everything went ok when the message say 'Plugin successfully activated'. Moving on with the next task they have, it's very easy to miss it or to forget about it shortly after.

Having that warning on the screens that functionality is not as expected, with the proper action to resolve the problem is a big help.

It also reduces the amount of 'actions' a user has to take to get the wanted functionality, one plugin does not have to be activated again after meeting all requirements.
And the error will disappear whenever all requirements are met. This means that if you install Yoast SEO you will still see the message that ACF needs to be installed as well, you don't need to activate the glue plugin again.

tldr;
Having errors in your face is not a bad thing, they will help the user to achieve the goal they have. It also reduces the number of actions they have to do to reach the goal.

File: inc/class-yoast-acf-analysis.php [hunk 5]
```diff
@@ -30,69 +30,65 @@ public function admin_init() {
...
-			);
-
-			if ( ! ($configuration instanceof Yoast_ACF_Analysis_Configuration) ) {
-				$configuration = $default_configuration;
-			}
-
-			Yoast_ACF_Analysis_Registry::instance()->add( 'config', $configuration );
-
+		$registry = Yoast_ACF_Analysis_Facade::get_registry();
+		if ( null !== $registry->get( 'config' ) ) {
```
Comment (moorscode) [inline inc/class-yoast-acf-analysis.php]: I read this message yesterday and let it sink in a bit more. Realizing that you come from an agency perspective the usecase to have this functionality is very clear.

Looking at the things that can be filtered, I would think that the specific modules would be the only thing that might change between projects, thus having to implement the entire interface to provide a custom configuration would be a lot of overhead.

What do you think about just filtering the class and thus providing the ability to extend the class instead of having to implement the entire interface?

File: inc/configuration/class-yoast-acf-analysis-configuration.php (hunk 1 above)
Comment (moorscode) [inline inc/configuration/class-yoast-acf-analysis-configuration.php]: Two reasons I prefer not to have the `__FUNCTION__` magic.
1. It's very unreadable for any developer to quickly understand what is going on and what the filter name is going to be
1. Whenever the method name changes the filter name would change as well, this is not something that would happen a lot but might introduce unexpected or overlooked behaviour

Comment (moorscode) [inline inc/configuration/class-yoast-acf-analysis-configuration.php]: I also prefer to name the methods to reflect what they are doing, thus `get_refresh_rate` would describe what is going on internally.

It's not just returning the refresh rate, it is building it up from dynamic sources.

**Comment by moorscode** [comment]
One of the things that need to be added is filter documentation, to make it easier to implement them for developers.

File: inc/class-yoast-acf-analysis-requirements.php (hunk 4 above)
Comment (kraftner) [inline inc/class-yoast-acf-analysis-requirements.php]: Okay, convinced. :smile:

File: inc/configuration/class-yoast-acf-analysis-configuration.php (hunk 1 above)
Comment (kraftner) [inline inc/configuration/class-yoast-acf-analysis-configuration.php]: Okay that was what I expected and I agree that plain strings improve readability. I was just curious if there are any other issues with `__FUNCTION__`I wasn't aware of.

File: inc/class-yoast-acf-analysis.php (hunk 5 above)
Comment (kraftner) [inline inc/class-yoast-acf-analysis.php]: I still think that an interface would be cleaner. I mean if you don't want to reimplement it from scratch you can still just extend the default implementation. But if you really want to you can still roll your own implementation. On the other hand I don't see what harm is done using an interface here.

Also the way to code is now you can still replace the configuration class with a custom implementation because there is [no check for the right class here](https://github.com/Yoast/yoast-acf-analysis/blob/292c0bc339e5c3544ba8c8def92a064ad85f6002/inc/class-yoast-acf-analysis.php#L48).

So this should probably be something like

`if ( null !== $registry->get( 'config' ) && $registry->get( 'config' ) instanceof Yoast_ACF_Analysis_Configuration ) {`
